test/harry/main/org/apache/cassandra/harry/model/BytesPartitionState.java

```diff
@@ -0,0 +1,554 @@
+            values = fromDescriptor(factory.staticPositions, rowState.vds);
+        }
+        else
+        {
+            clustering = factory.clusteringCache.inflate(rowState.cd);
+            values = fromDescriptor(factory.regularPositions, rowState.vds);
+        }
+        return new Row(clustering, values);
+    }
+
+    public Collection<Row> rows()
+    {
+        return state.rows().values().stream().map(this::toRow).collect(Collectors.toList());
+    }
+
+    public NavigableSet<Clustering<ByteBuffer>> clusteringKeys()
+    {
+        NavigableSet<Clustering<ByteBuffer>> navigableSet = new TreeSet<>(factory.clusteringComparator);
+        state.rows().keySet().stream().map(factory.clusteringCache::inflate).forEach(navigableSet::add);
+        return navigableSet;
+    }
+
+    public Row staticRow()
+    {
+        return toRow(state.staticRow());
+    }
+
+    public boolean shouldDelete()
+    {
+        return state.shouldDelete();
+    }
+
+    static List<String> asCQL(ImmutableUniqueList<Symbol> columns, ByteBuffer[] row)
+    {
+        List<String> cql = new ArrayList<>(row.length);
+        for (int i = 0; i < row.length; i++)
+        {
+            AbstractType type = columns.get(i).type();
+            cql.add(type.toCQLString(row[i]));
+        }
+        return cql;
+    }
+
+    public class PrimaryKey implements Comparable<PrimaryKey>
+    {
+        public final BytesPartitionState.Ref partition;
+        @Nullable
+        public final Clustering<ByteBuffer> clustering;
+
+        public PrimaryKey(BytesPartitionState.Ref partition, @Nullable Clustering<ByteBuffer> clustering)
+        {
+            this.partition = partition;
+            this.clustering = clustering;
+        }
+
+        @Override
+        public int compareTo(PrimaryKey o)
+        {
+            int rc = partition.compareTo(o.partition);
+            if (rc != 0) return rc;
+            if (clustering == null) return rc; // if the partition matches, and clustering is null (partition doesn't have rows) then it would be a bug if o.clustering was non-null
+            rc = factory.clusteringComparator.compare(clustering, o.clustering);
+            return rc;
+        }
+
+        @Override
+        public boolean equals(Object o)
+        {
+            if (this == o) return true;
+            if (o == null || getClass() != o.getClass()) return false;
+            PrimaryKey that = (PrimaryKey) o;
+            return partition.equals(that.partition) && Objects.equals(clustering, that.clustering);
+        }
+
+        @Override
+        public int hashCode()
+        {
+            return Objects.hash(partition, clustering);
+        }
+    }
+
+    public static class Ref implements Comparable<Ref>
+    {
+        private final Factory factory;
+        @Nullable
+        public final Clustering<ByteBuffer> key;
+        public final Token token;
+        private final boolean nullKeyGtMatchingToken;
```

Comment:
Comment explaining how this is used to deal with multiple rows with the same token and head/tail sets.

Reply:
this is good feedback... its been a pain to get this right and i forget it 10s later...